### PR TITLE
batched inference

### PR DESCRIPTION
**Pull request by varshith15**
- [x] batched inference llama unit testcase
- [x] batched inference llava unit testcase
- [x] take list as input for prompt, image_str, request_id
- [x] fix callbacks order
- [x] online batching in chatgpt api
- [x] top_p!=1 for batched patch
- [x] early stopping in the batch if the stop has eos occurred before, currently only check the last token
- [x] online batching conditions update (timeout, max_batch etc)

### CODE REVIEW

**Comment by varshith15** [comment]
hey @AlexCheema 
to start with, i am assuming that inflight batching isnt needed right away
we can just batch requests in the api section and do inference and split them back in the api section thoughts?

**Comment by AlexCheema** [comment]
> hey @AlexCheema to start with, i am assuming that inflight batching isnt needed right away (have to revamp a lot the api to make inflight happen) we can just batch requests in the api section and do inference and split them back in the api section thoughts?

Sounds good. Yes, no need for inflight batching - we can add that in a subsequent PR.
Lets do this!

Related issue: https://github.com/exo-explore/exo/issues/1

**Comment by varshith15** [comment]
Hey @AlexCheema 
I was just working on the batching of requests at the endpoint part but it just seems a little gimmck-y to merge them and separate them after processing because there are a bunch of moving parts like stream functionality etc

My current idea is to basically combine the requests ("req1_req2_req3..") and then check if req_id is present in the broadcasted result id and split it based on the position of the id. 

do you have a better idea? what did you have in your mind?

**Comment by varshith15** [comment]
ive thought of a better way, just need to update functions to take list of req_ids, prompts and img_strs and broadcast back the specific results based on the index of the req in the back - that way dont have to change a lot of code in the chatgpt api

**Comment by varshith15** [comment]
hey @AlexCheema, batched inference with online batching works now. There are few more small patches required, ill push them in a bit. Could you please review the idea meanwhile.

**Comment by varshith15** [comment]
@AlexCheema it's done! PRM

**Comment by AlexCheema** [comment]
Thanks @varshith15 I will take a proper look at some point next week.

**Comment by AlexCheema** [comment]
I haven't had a proper look yet, but I'm wondering what the behaviour is when a request is being processed and another request comes in?

**Comment by varshith15** [comment]
currently the requests just keep getting processed one after the other, by processes i mean after the await process_prompt is done but maybe a semaphore is needed

**Comment by AlexCheema** [comment]
> currently the requests just keep getting processed one after the other, by processes i mean after the await process_prompt is done but maybe a semaphore is needed

I think this might cause issues currently since the kv_cache is shared. Can you test this out to see if it works? It doesn't work on main but maybe with your changes it works now.

**Comment by varshith15** [comment]
I've tested it out in a bunch of different scenarios with sending the reqs to the service, it works as expected

**Comment by varshith15** [comment]
@AlexCheema you are right, there a bug when theres a request running already and a new request comes, the older one is not able to return, checking it

**Comment by varshith15** [comment]
@AlexCheema could you expand on why concurrent processing doesn't work as the kv_csche is shared?

**Comment by AlexCheema** [comment]
> @AlexCheema could you expand on why concurrent processing doesn't work as the kv_csche is shared?
> 
> I haven't debugged yet, will do it over the weekend

I pushed something for MLX that uses a LRU cache with a bunch of KV caches that should fix it. Can you merge the latest exo and fix conflicts then I will take a look through the PR.

**Comment by varshith15** [comment]
@AlexCheema still doesnt work as expected, i tried sending two reqs with a slight delay between them, one gives the expected answer and the other gives some unicode
checking this

**Comment by varshith15** [comment]
i think we should have a semaphore and have the max_caches value to be equal to semaphore
with the LRU cache fix the previous error is fixed now, it works as expected

File: exo/helpers.py [hunk 1]
```diff
@@ -93,22 +94,30 @@ def terminal_link(uri, label=None):
...
+
+      await asyncio.wait_for(wait_for_valid_result(), timeout)
+      return self.result.popleft()
 
   def on_next(self, callback: Callable[..., None]) -> None:
     self.observers.append(callback)
 
   def set(self, *args: T) -> None:
-    self.result = args
+    self.result.append(args)
```
Comment (varshith15) [inline exo/helpers.py:117]: @AlexCheema  the concurrent execution issue is due to this, not due to the LRU

**Comment by varshith15** [comment]
@AlexCheema  this is done now, please review

**Pull request closed without merging**